--- a/src/audio/decode/decode_alsa_backend.c
+++ b/src/audio/decode/decode_alsa_backend.c
@@ -715,8 +715,9 @@
 		return err;
 	}
 
-	/* set the sample format */
-	fmt = 0;
+	/* set the sample format, force S16_LE for effects stream */
+	if (state->flags & FLAG_STREAM_EFFECTS) {fmt = 3;} else {fmt = 0;}
+
 	while (fmts[fmt] != SND_PCM_FORMAT_UNKNOWN) {
 		if (snd_pcm_hw_params_set_format(*pcmp, hw_params, fmts[fmt]) >= 0) {
 			LOG_INFO("Opened device %s using format: %s sample rate: %u", device, snd_pcm_format_name(fmts[fmt]), sample_rate);
